service/windows/installer/custom-actions/src/lib.rs

```diff
@@ -0,0 +1,173 @@
+use std::ffi::{OsStr, OsString};
+use std::fs::remove_dir_all;
+use std::os::windows::ffi::{OsStrExt, OsStringExt};
+use std::path::Path;
+use std::ptr;
+use winapi::shared::minwindef::{DWORD, LPDWORD, UINT};
+use winapi::shared::winerror::{ERROR_FUNCTION_NOT_CALLED, ERROR_MORE_DATA, ERROR_SUCCESS};
+use winapi::um::winnt::{LPCWSTR, LPWSTR};
+use winapi::um::winsvc::*;
+
+type MsiHandle = u32;
+
+extern "system" {
+    fn MsiGetPropertyW(
+        hInstall: MsiHandle,
+        szName: LPCWSTR,
+        szValueBuf: LPWSTR,
+        pcchValueBuf: LPDWORD,
+    ) -> UINT;
+}
+
+/// A custom action to delete a directory and all of its contents.
+#[no_mangle]
+pub extern "system" fn DeleteDataFolder(install_handle: MsiHandle) -> UINT {
+    match get_data(install_handle) {
+        Err(_) => {
+            return ERROR_FUNCTION_NOT_CALLED;
+        }
+        Ok(data_dir_wide) => {
+            let data_dir = OsString::from_wide(&data_dir_wide[0..data_dir_wide.len() - 1]);
+            let p = Path::new(&data_dir);
+            if remove_dir_all(p).is_err() {
+                return ERROR_FUNCTION_NOT_CALLED;
+            }
+        }
+    }
+    ERROR_SUCCESS
+}
+
+const SERVICE_AUTO_START: DWORD = 0x00000002;
+const SERVICE_DEMAND_START: DWORD = 0x00000003;
```

Comment:
What is the meaning of these? Can there be documentation?

Reply:
Added